src/os/filestore/FileStore.cc

```diff
@@ -3536,17 +3536,30 @@ int FileStore::_do_sparse_copy_range(int from, int to, uint64_t srcoff, uint64_t
   if (len == 0)
     return 0;
 
-  if (backend->has_seek_data_hole()) {
+  int64_t written = 0;
+  struct stat st;
+  r = ::fstat(to, &st);
+  if (r < 0) {
+    r = -errno;
+    derr << __func__ << ": fstat error at " << to << " " << cpp_strerror(r) << dendl;
+    goto out;
+  }
+
+  if ((uint64_t)st.st_size > dstoff) {
```

Comment:
I don't think this condition is correct.

Example: 
- `from` is 8KB long, and first 4KB is a hole
- `to` is 1MB long and has no holes
- srcoff = 0, len = 4096, dstoff = 1MB - 4KB
Such operation should have punched a 4KB hole at 1MB - 4K, however nothing will happen since the (st.size > dstoff) condition does not hold.

The problem is that this method only copies the data, however it should have also punched the same holes as in the source.